Allow DDP to wrap multi-GPU modules

mrshenli opened the pull request:
Summary: allow DDP to take multi-gpu models

Differential Revision: D14822375

pietern commented on torch/nn/parallel/distributed.py:
```diff
@@ -153,11 +153,22 @@ class DistributedDataParallel(Module):
         module (Module): module to be parallelized
-        device_ids (list of int or torch.device): CUDA devices (default: all devices)
-        output_device (int or torch.device): device location of output (default: device_ids[0])
+        device_ids (list of int or torch.device):  CUDA devices. This should
```
Extra whitespace -- is this intentional?

mrshenli replied:
no, not intentional. I will edit, thanks!

pietern reviewed (approved):
Some minor points. Looking good overall. Glad that we'll be able to support multi device modules here!

mrshenli commented on test/test_c10d.py:
```diff
@@ -1527,11 +1581,43 @@ def _test_ddp_with_process_group(self, process_group, gpus):
+            model = QuadraGpuNet(gpus)
+
+        ddp_model = DistributedDataParallel(
+            copy.deepcopy(model),
```
Don't we need to make sure that `model` and `ddp_model` operate on independent params so that we can compare?

pietern replied:
Never mind, it's needed because of the numerical equivalence testing.

mrshenli committed:
Allow DDP to wrap multi-GPU modules (#19271)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/19271

allow DDP to take multi-gpu models

Reviewed By: pietern

Differential Revision: D14822375

fbshipit-source-id: 8c8bcd4526643be5fa44134620d58fcf2c197238
